File: 20.valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> stk;
        for (auto c: s) {
            if (c == '(' || c == '{' || c == '[') {
                stk.push(c);
            } else if (c == ')') {
                if (!stk.empty() && stk.top() == '(') {
                    stk.pop();
                } else {
                    return false;
                }
            } else if (c == ']') {
                if (!stk.empty() && stk.top() == '[') {
                    stk.pop();
                } else {
                    return false;
                }
            } else if (c == '}') {
                if (!stk.empty() && stk.top() == '{') {
                    stk.pop();
                } else {
                    return false;
                }
            }
        }
        return stk.empty();
    }
};

File: 20.valid-parentheses.py
class Solution:
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stk = []
        for c in s:
            if c == '(' or c == '{' or c == '[':
                stk.append(c);
            elif c == ')':
                if len(stk) == 0 or stk.pop() != '(':
                    return False
            elif c == ']':
                if len(stk) == 0 or stk.pop() != '[':
                    return False
            elif c == '}':
                if len(stk) == 0 or stk.pop() != '{':
                    return False
        return len(stk) == 0